src/interface/mapping/vec.rs

```diff
@@ -0,0 +1,223 @@
+// This file is part of Astarte.
+//
+// Copyright 2023 SECO Mind Srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Sorted collection of interfaces accessible by endpoint or path.
+
+use std::{
+    cmp::Ordering, collections::BTreeSet, fmt::Debug, iter::FusedIterator, ops::Deref,
+    slice::Iter as SliceIter,
+};
+
+use itertools::Itertools;
+
+use super::{endpoint::Endpoint, InterfaceMapping};
+
+#[derive(Debug, Clone, Default)]
+pub(crate) struct MappingVec<T> {
+    mappings: Vec<Item<T>>,
+}
+
+impl<T> MappingVec<T> {
+    /// Create and empty vector.
+    pub(crate) fn new() -> Self {
+        Self {
+            mappings: Vec::new(),
+        }
+    }
+
+    /// Get's the mapping searching the [`Endpoint`]'s for the one matching the path.
+    pub(crate) fn get<U>(&self, path: &U) -> Option<&T>
+    where
+        T: InterfaceMapping + Debug,
+        U: PartialOrd<Endpoint<String>> + Eq + Debug,
+    {
+        self.mappings
+            .binary_search_by(|item| {
+                let endpoint = item.endpoint();
+                // We invert the ordering for less/greater since the trait is implemented on the
+                // path, but we need to actually check the argument
+                if path.eq(endpoint) {
```

Comment:
Even if this is practically the same, it's not fully correct. Since you're trying to order a PartialOrd sequence of items (eg. floats which have Nan) and giving them all the out of domain values (+Nan and -Nan that are out of R) the Ordering::Less. I would like to keep the comparison explicit like this, even if in practice it will never happen. Here is an example https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=501d604dd6db6c4ccd0414f79860fc7c